README.md
What?
=====

A simple Graphite dashboard built using Twitters Bootstrap.

Adding new dashboards is very easy and individual graphs is
described using a small DSL.

See the _sample_ directory for an example dashboard including
a screenshot.

Config?
-------

This dashboard is a Sinatra application, I suggest deploying it
in Passenger or other Sinatra application server.

A sample _gdash.yaml-sample_ is included, you should rename it to
_gdash.yaml_ and adjust the url to your Graphite etc in there.

The SinatraApp class take two required arguments:

    * Where graphite is installed
    * The directory that has your _dashboards_ directory full of templates

and additional options:

    * The title to show at the top of your Graphite
    * A prefix to prepend to all URLs in the dashboard
    * How many columns of graphs to create, 2 by default.
    * How often dashboard page is refreshed, 60 sec by default.
    * The width of the graphs, 500 by default
    * The height of the graphs, 250 by default
    * Where your whisper files are stored - future use
    * Optional interval quick filters

Creating Dashboards?
--------------------

You can have multiple top level categories of dashboard.  Just create directories
in the _templatedir_ for each top level category.

In each top level category create a sub directory with a short name for each new dashboard.

You need a file called _dash.yaml_ for each dashboard, here is a sample:

    :name: Email Metrics
    :description: Hourly metrics for the email system

Then create descriptions in files like _cpu.graph_ in the same directory, here
is a sample:

    title       "Combined CPU Usage"
    vtitle      "percent"
    area        :stacked
    description "The combined CPU usage for all Exim Anti Spam servers"

    field :iowait, :scale => 0.001,
                   :color => "red",
                   :alias => "IO Wait",
                   :data  => "sumSeries(derivative(mw*munin.cpu.iowait))"

    field :system, :scale => 0.001,
                   :color => "orange",
                   :alias => "System",
                   :data  => "sumSeries(derivative(mw*.munin.cpu.system))"

    field :user, :scale => 0.001,
                 :color => "yellow",
                 :alias => "User",
                 :data  => "sumSeries(derivative(mw*.munin.cpu.user))"

The dashboard will use the _description_ field to show popup information bubbles
when someone hovers over a graph with their mouse for 2 seconds.

The graphs are described using a DSL that has its own project and documented
over at https://github.com/ripienaar/graphite-graph-dsl/wiki

At the moment we do not support the _Related Items_ feature of the DSL.

Template Directory Layout?
--------------------------

The directory layout is such that you can have many groupins of dashboards each with
many dashboards underneath it, an example layout of your templates dir would be:

	graph_templates
	`-- virtualization
	    |-- dom0
	    |   |-- dash.yaml
	    |   |-- iowait.graph
	    |   |-- load.graph
	    |   |-- system.graph
	    |   |-- threads.graph
	    |   `-- user.graph
	    `-- kvm1
		|-- dash.yaml
		|-- disk_read.graph
		|-- disk_write.graph
		|-- ssd_read.graph
		`-- ssd_write.graph

Here we have a group of dashboards called 'virtualization' with 2 dashboards inside it
each with numerous graphs.

You can create as many groups as you want each with many dashboards inside.

Custom Time Intervals?
--------------------

You can reuse your dashboards and adjust the time interval by using the following url
structure:

    http://gdash.example.com/dashboard/email/time/-8d/-7d

or

    http://gdash.example.com/dashboard/email/?from=-8d&until=-7d
    http://gdash.example.com/dashboard/email/full/2/600/300?from=-8d&until=-7d

This will display the _email_ dashboard with a time interval same day last week.
If you hit */dashboard/email/time/* it will default to the past hour (*-1hour*)
See http://graphite.readthedocs.org/en/1.0/url-api.html#from-until for more info
acceptable *from* and *until* values.

Quick interval filters shown in interface are configurable in _gdash.yaml_ options sections. Eg:

	:options:
           :interval_filters:
             - :label: Last Hour
               :from: -1h
               :to: now
             - :label: Last Day
               :from: -1day
             - :label: Current Week
               :from: monday
               :to: now

Quick filter is not shown when *interval_filters* section is missing in configuration file.

Time Intervals Display?
-----------------------

If you configure time intervals in the config file you can click on any graph in
the main dashboard view and get a view with different time intervals of the same
graph

	:options:
	  :intervals:
	    - [ "-1hour", "1 hour" ]
	    - [ "-2hour", "2 hour" ]
	    - [ "-1day", "1 day" ]
	    - [ "-1month", "1 month" ]
	    - [ "-1year", "1 year" ]

With this in place in the _config.yaml_ clicking on a graph will show the 5 intervals
defined above of that graph

Full Screen Displays?
---------------------

You can reuse your dashboards for big displays against a wall in your NOC or office
by using the following url structure:

    http://gdash.example.com/dashboard/email/full/4/600/300
    http://gdash.example.com/dashboard/email/full/4?width=600&height=300

This will display the _email_ dashboard in _4_ columns each graph with a width of
_600_ and a height of _300_

The screen will refresh every minute

Contact?
--------

R.I.Pienaar / rip@devco.net / http://www.devco.net/ / @ripienaar


CHANGELOG
2012/02/11 - Support time interval views for any graph
2011/11/12 - Include the uncompiled less files and a js script to compile on demand
             making it easier to customize the look and feel of gdash - thanks Joe Miller
2011/10/18 - Multiple top level categories are supported
2011/10/16 - The settings that used to be in config.ru is now in a config file
2011/10/10 - Add a full screen dashboard mode


lib/graphite_graph.rb
require 'uri'
# A small DSL to assist in the creation of Graphite graphs
# see https://github.com/ripienaar/graphite-graph-dsl/wiki
# for full details
class GraphiteGraph
  attr_reader :info, :properties, :targets, :target_order, :critical_threshold, :warning_threshold

  def initialize(file, overrides={}, info={})
    @info = info
    @file = file
    @munin_mode = false
    @overrides = overrides
    @linecount = 0

    @critical_threshold = []
    @warning_threshold = []

    load_graph
  end

  def defaults
    @properties = {:title => nil,
                   :vtitle => nil,
                   :width => 500,
                   :height => 250,
                   :from => "-1hour",
                   :until => "now",
                   :surpress => false,
                   :description => nil,
                   :hide_legend => nil,
                   :hide_grid => nil,
                   :ymin => nil,
                   :ymax => nil,
                   :linewidth => nil,
                   :linemode => nil,
                   :fontsize => nil,
                   :fontbold => false,
                   :timezone => nil,
                   :draw_null_as_zero => false,
                   :major_grid_line_color => nil,
                   :minor_grid_line_color => nil,
                   :area => :none}.merge(@overrides)
  end

  def [](key)
    if key == :url
      url
    else
      @properties[key]
    end
  end

  def method_missing(meth, *args)
    if properties.include?(meth)
      properties[meth] = args.first unless @overrides.include?(meth)
    else
      super
    end
  end

  def load_graph
    @properties = defaults
    @targets = {}
    @target_order = []

    self.instance_eval(File.read(@file)) unless @file == :none
  end

  def service(service, data, &blk)
    raise "No hostname given for this instance" unless info[:hostname]

    @service_mode = {:service => service, :data => data}

    blk.call

    @service_mode = false
  end

  # add forecast, bands, aberrations and actual fields using the
  # Holt-Winters Confidence Band prediction model
  #
  #    hw_predict :foo, :data => "some.data.item", :alias => "Some Item"
  #
  # You can tweak the colors by setting:
  #     :forecast_color => "blue"
  #     :bands_color => "grey"
  #     :aberration_color => "red"
  #
  # You can add an aberration line:
  #
  #     :aberration_line => true,
  #     :aberration_second_y => true
  #
  # You can disable the forecast line by setting:
  #
  #     :forecast_line => false
  #
  # You can disable the confidence lines by settings:
  #
  #     :bands_lines => false
  #
  # You can disable the display of the actual data:
  #
  #     :actual_line => false
  def hw_predict(name, args)
    raise ":data is needed as an argument to a Holt-Winters Confidence forecast" unless args[:data]

    unless args[:forecast_line] == false
      forecast_args = args.clone
      forecast_args[:data] = "holtWintersForecast(#{forecast_args[:data]})"
      forecast_args[:alias] = "#{args[:alias]} Forecast"
      forecast_args[:color] = args[:forecast_color] || "blue"
      field "#{name}_forecast", forecast_args
    end

    unless args[:bands_lines] == false
      bands_args = args.clone
      bands_args[:data] = "holtWintersConfidenceBands(#{bands_args[:data]})"
      bands_args[:color] = args[:bands_color] || "grey"
      bands_args[:dashed] = true
      bands_args[:alias] = "#{args[:alias]} Confidence"
      field "#{name}_bands", bands_args
    end

    if args[:aberration_line]
      aberration_args = args.clone
      aberration_args[:data] = "holtWintersAberration(keepLastValue(#{aberration_args[:data]}))"
      aberration_args[:color] = args[:aberration_color] || "orange"
      aberration_args[:alias] = "#{args[:alias]} Aberration"
      aberration_args[:second_y_axis] = true if aberration_args[:aberration_second_y]
      field "#{name}_aberration", aberration_args
    end

    if args[:critical]
      color = args[:critical_color] || "red"
      critical :value => args[:critical], :color => color, :name => name
    end

    if args[:warning]
      color = args[:warning_color] || "orange"
      warning :value => args[:warning], :color => color, :name => name
    end

    args[:color] ||= "yellow"

    field name, args unless args[:actual_line] == false
  end

  alias :forecast :hw_predict

  # takes a series of metrics in a wildcard query and aggregates the values by a subgroup
  #
  # data must contain a wildcard query, a subgroup position, and an optional aggregate function.
  # if the aggregate function is omitted, sumSeries will be used.
  #
  # group :data => "metric.*.value", :subgroup => "2", :aggregator => "sumSeries"
  #
  def group(name, args)
    raise ":data is needed as an argument to group metrics" unless args[:data]
    raise ":subgroup is needed as an argument to group metrics" unless args.include?(:subgroup)

    args[:aggregator] = "sumSeries" unless args[:aggregator]

    group_args = args.clone
    group_args[:data] = "groupByNode(#{group_args[:data]},#{group_args[:subgroup]},\"#{group_args[:aggregator]}\")"
    field "#{name}_group", group_args

  end

  # draws a single dashed line with predictable names, defaults to red line
  #
  # data can be a single item or a 2 item array, it doesn't break if you supply
  # more but # more than 2 items just doesn't make sense generally
  #
  # critical :value => [700, -700], :color => "red"
  #
  # You can prevent the line from being drawn but just store the ranges for monitoring
  # purposes by adding :hide => true to the arguments
  def critical(options)
    raise "critical lines need a value" unless options[:value]

    @critical_threshold = [options[:value]].flatten

    options[:color] ||= "red"

    unless options[:hide]
      @critical_threshold.each_with_index do |crit, index|
        line :caption => "crit_#{index}", :value => crit, :color => options[:color], :dashed => true
      end
    end
  end

  # draws a single dashed line with predictable names, defaults to orange line
  #
  # data can be a single item or a 2 item array, it doesn't break if you supply
  # more but # more than 2 items just doesn't make sense generally
  #
  # warning :value => [700, -700], :color => "orange"
  #
  # You can prevent the line from being drawn but just store the ranges for monitoring
  # purposes by adding :hide => true to the arguments
  def warning(options)
    raise "warning lines need a value" unless options[:value]

    @warning_threshold = [options[:value]].flatten

    options[:color] ||= "orange"

    unless options[:hide]
      @warning_threshold.flatten.each_with_index do |warn, index|
        line :caption => "warn_#{index}", :value => warn, :color => options[:color], :dashed => true
      end
    end
  end

  # draws a simple line on the graph with a caption, value and color.
  #
  # line :caption => "warning", :value => 50, :color => "orange"
  def line(options)
    raise "lines need a caption" unless options.include?(:caption)
    raise "lines need a value" unless options.include?(:value)
    raise "lines need a color" unless options.include?(:color)

    options[:alias] = options[:caption] unless options[:alias]

    args = {:data => "threshold(#{options[:value]})", :color => options[:color], :alias => options[:alias]}

    args[:dashed] = true if options[:dashed]

    field "line_#{@linecount}", args

    @linecount += 1
  end

  # adds a field to the graph, each field needs a unique name
  def field(name, args)
    raise "A field called #{name} already exist for this graph" if targets.include?(name)

    default = {}

    if @service_mode
      default[:data] = [info[:hostname], @service_mode[:service], @service_mode[:data], name].join(".")
    end

    targets[name] = default.merge(args)
    target_order << name
  end

  def url(format = nil, url=true)
    return nil if properties[:surpress]

    url_parts = []
    colors = []

    [:title, :vtitle, :from, :width, :height, :until].each do |item|
      url_parts << "#{item}=#{properties[item]}" if properties[item]
    end

    url_parts << "areaMode=#{properties[:area]}" if properties[:area]
    url_parts << "hideLegend=#{properties[:hide_legend]}" if properties.include?(:hide_legend)
    url_parts << "hideGrid=#{properties[:hide_grid]}" if properties[:hide_grid]
    url_parts << "yMin=#{properties[:ymin]}" if properties[:ymin]
    url_parts << "yMax=#{properties[:ymax]}" if properties[:ymax]
    url_parts << "lineWidth=#{properties[:linewidth]}" if properties[:linewidth]
    url_parts << "lineMode=#{properties[:linemode]}" if properties[:linemode]
    url_parts << "fontSize=#{properties[:fontsize]}" if properties[:fontsize]
    url_parts << "fontBold=#{properties[:fontbold]}" if properties[:fontbold]
    url_parts << "drawNullAsZero=#{properties[:draw_null_as_zero]}" if properties[:draw_null_as_zero]
    url_parts << "tz=#{properties[:timezone]}" if properties[:timezone]
    url_parts << "majorGridLineColor=#{properties[:major_grid_line_color]}" if properties[:major_grid_line_color]
    url_parts << "minorGridLineColor=#{properties[:minor_grid_line_color]}" if properties[:minor_grid_line_color]

    target_order.each do |name|
      target = targets[name]

      if target[:target]
        url_parts << "target=#{target[:target]}"
      else
        raise "field #{name} does not have any data associated with it" unless target[:data]

        graphite_target = target[:data]

        graphite_target = "derivative(#{graphite_target})" if target[:derivative]
        graphite_target = "scale(#{graphite_target},#{target[:scale]})" if target[:scale]
        graphite_target = "drawAsInfinite(#{graphite_target})" if target[:line]
        graphite_target = "movingAverage(#{graphite_target},#{target[:smoothing]})" if target[:smoothing]

        graphite_target = "color(#{graphite_target},\"#{target[:color]}\")" if target[:color]
        graphite_target = "dashed(#{graphite_target})" if target[:dashed]
        graphite_target = "secondYAxis(#{graphite_target})" if target[:second_y_axis]

        unless target.include?(:subgroup)
          if target[:alias]
            if target[:alias] == "none"
            	graphite_target = graphite_target
            else
            	graphite_target = "alias(#{graphite_target},\"#{target[:alias]}\")"
             end
          else
            graphite_target = "alias(#{graphite_target},\"#{name.to_s.capitalize}\")"
          end
        end

        url_parts << "target=#{graphite_target}"
      end
    end

    url_parts << "format=#{format}" if format

    if url
       URI.encode(url_parts.join("&"))
    else
      url_parts
    end
  end
end
